hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/FrameUtils.java

```diff
@@ -257,25 +257,47 @@ public static boolean stackIncludesActiveAddress(
     public enum CallType {
         QUALIFIED_DELEGATE,
         UNQUALIFIED_DELEGATE,
-        DIRECT_OR_TOKEN_REDIRECT,
+        DIRECT_OR_PROXY_REDIRECT,
     }
 
     public static CallType callTypeOf(final MessageFrame frame) {
         if (!isDelegateCall(frame)) {
-            return CallType.DIRECT_OR_TOKEN_REDIRECT;
+            return CallType.DIRECT_OR_PROXY_REDIRECT;
         }
         final var recipient = frame.getRecipientAddress();
         // Evaluate whether the recipient is either a token or on the permitted callers list.
         // This determines if we should treat this as a delegate call.
         // We accept delegates if the token redirect contract calls us.
         final CallType viableType;
         if (isToken(frame, recipient)) {
-            viableType = CallType.DIRECT_OR_TOKEN_REDIRECT;
+            viableType = CallType.DIRECT_OR_PROXY_REDIRECT;
         } else if (isQualifiedDelegate(recipient, frame)) {
             viableType = CallType.QUALIFIED_DELEGATE;
         } else {
             return CallType.UNQUALIFIED_DELEGATE;
         }
+        // make sure we have a parent calling context
+        return parentIsNonDelegate(frame, viableType);
+    }
+
+    public static CallType callTypeForAccountOf(final MessageFrame frame) {
+        if (!isDelegateCall(frame)) {
+            return CallType.DIRECT_OR_PROXY_REDIRECT;
+        }
+        final var recipient = frame.getRecipientAddress();
+        // Evaluate whether the recipient is a regular account.
+        // This determines if we should treat this as a delegate call.
+        // We accept delegates if the account redirect contract calls us.
+        final CallType viableType;
+        if (isRegularAccount(frame, recipient)) {
+            viableType = CallType.DIRECT_OR_PROXY_REDIRECT;
+        } else {
+            return CallType.UNQUALIFIED_DELEGATE;
+        }
+        return parentIsNonDelegate(frame, viableType);
```

Comment:
why not move L297 up to after L293?

Reply:
Also `parentIsNonDelegate` may be the wrong name - kind of looks like a predicate which it's not.